File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/merge_sort.py
"""Merge sort."""


def sort(values: list[int]) -> list[int]:
    """Return a sorted copy of the input list."""
    if values is None:
        raise ValueError("values must not be None")
    if len(values) <= 1:
        return list(values)

    mid = len(values) // 2
    left = sort(values[:mid])
    right = sort(values[mid:])
    return _merge(left, right)


def _merge(left: list[int], right: list[int]) -> list[int]:
    result: list[int] = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] <= right[right_index]:
            result.append(left[left_index])
            left_index += 1
        else:
            result.append(right[right_index])
            right_index += 1

    result.extend(left[left_index:])
    result.extend(right[right_index:])
    return result

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/MergeSort.java
package com.divideconquer.algorithms;

import java.util.Arrays;

public final class MergeSort {
    private MergeSort() {
    }

    public static int[] sort(int[] values) {
        if (values == null) {
            throw new IllegalArgumentException("values must not be null");
        }
        if (values.length <= 1) {
            return Arrays.copyOf(values, values.length);
        }

        int mid = values.length / 2;
        int[] left = sort(Arrays.copyOfRange(values, 0, mid));
        int[] right = sort(Arrays.copyOfRange(values, mid, values.length));
        return merge(left, right);
    }

    private static int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int leftIndex = 0;
        int rightIndex = 0;
        int resultIndex = 0;

        while (leftIndex < left.length && rightIndex < right.length) {
            if (left[leftIndex] <= right[rightIndex]) {
                result[resultIndex++] = left[leftIndex++];
            } else {
                result[resultIndex++] = right[rightIndex++];
            }
        }
        while (leftIndex < left.length) {
            result[resultIndex++] = left[leftIndex++];
        }
        while (rightIndex < right.length) {
            result[resultIndex++] = right[rightIndex++];
        }
        return result;
    }
}
